toolchains/prebuilt_toolchains.py

```diff
@@ -0,0 +1,370 @@
+#!/usr/bin/env python3
+
+from pathlib import Path
+from textwrap import indent
+import urllib.request
+import hashlib
+
+
+CMAKE_SHA256_URL_TEMPLATE = "https://cmake.org/files/v{minor}/cmake-{full}-SHA-256.txt"
+CMAKE_URL_TEMPLATE = "https://github.com/Kitware/CMake/releases/download/v{full}/{file}"
+
+CMAKE_VERSIONS = [
+    "3.19.4",
+    "3.18.6",
+    "3.17.5",
+    "3.16.9",
+    "3.15.7",
+    "3.14.7",
+]
+
+CMAKE_TARGETS = {
+    "Darwin-x86_64": [
+        "@platforms//cpu:x86_64",
+        "@platforms//os:macos",
+    ],
+    "Linux-aarch64": [
+        "@platforms//cpu:aarch64",
+        "@platforms//os:linux",
+    ],
+    "Linux-x86_64": [
+        "@platforms//cpu:x86_64",
+        "@platforms//os:linux",
+    ],
+    "macos-universal": [
+        "@platforms//os:macos",
+    ],
+    "win32-x86": [
+        "@platforms//cpu:x86_32",
+        "@platforms//os:windows",
+    ],
+    "win64-x64": [
+        "@platforms//cpu:x86_64",
+        "@platforms//os:windows",
+    ],
+}
+
+NINJA_URL_TEMPLATE = "https://github.com/ninja-build/ninja/releases/download/v{full}/ninja-{target}.zip"
+
+NINJA_TARGETS = {
+    "linux": [
+        "@platforms//os:linux",
+    ],
+    "mac": [
+        "@platforms//os:macos",
+    ],
+    "win": [
+        "@platforms//os:windows",
+    ],
+}
+
+NINJA_VERSIONS = (
+    "1.10.2",
+    "1.10.1",
+    "1.10.0",
+    "1.9.0",
+    "1.8.2",
+)
+
+REPO_DEFINITION = """\
+maybe(
+    http_archive,
+    name = "{name}",
+    urls = [
+        "{url}",
+    ],
+    sha256 = "{sha256}",
+    strip_prefix = "{prefix}",
+    build_file_content = {template}.format(
+        bin = "{bin}",
+        exec_compatible_with = "{exec_compatible_with}",
+    ),
+)
+"""
+
+REGISTER_TOOLCHAINS = """\
+native.register_toolchains(
+{toolchains}
+)
+"""
+
+BZL_FILE_TEMPLATE = """\
+\"\"\" A U T O G E N E R A T E D  -- D O   N O T   M O D I F Y
+@generated
+
+This file is generated by prebuilt_toolchains.py
+\"\"\"
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
+
+_CMAKE_BUILD_FILE = \"\"\"\
+load("@rules_foreign_cc//tools/build_defs/native_tools:native_tools_toolchain.bzl", "native_tool_toolchain")
+
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "cmake_data",
+    srcs = glob(
+        [
+            "**",
+        ],
+        exclude = [
+            "WORKSPACE",
+            "WORKSPACE.bazel",
+            "BUILD",
+            "BUILD.bazel",
+        ],
+    ),
+)
+
+filegroup(
+    name = "cmake_bin",
+    srcs = ["bin/{{bin}}"],
+    data = [":cmake_data"],
+)
+
+native_tool_toolchain(
+    name = "cmake_tool",
+    path = "$(execpath :cmake_bin)",
+    target = ":cmake_bin",
+)
+
+toolchain(
+    name = "cmake_toolchain",
+    exec_compatible_with = {{exec_compatible_with}},
+    toolchain = ":cmake_tool",
+    toolchain_type = "@rules_foreign_cc//tools/build_defs:cmake_toolchain",
+)
```

Comment:
The issue with having the toolchain definition inside the repository is that this causes bazel to fetch tools for all platforms on all platforms. Can this be moved into `@rules_foreign_cc` as I have done in #507?

Reply:
Good call, updated.